src/compare.ts

```diff
@@ -0,0 +1,173 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+/**
+ * Small wrapper for lazily producing values.
+ * @template T the value type.
+ */
+class Lazy<T> {
+  private calculatedValue?: T;
+  private hasCalculatedValue: boolean = false;
+
+  /**
+   * Returns the value or calculates it if none has yet been calculated.
+   * @returns {T} the value.
+   */
+  get value(): T {
+    if (!this.hasCalculatedValue) {
+      this.calculatedValue = this.calculate();
+      this.hasCalculatedValue = true;
+    }
+    return <T>this.calculatedValue;
+  }
+
+  /**
+   * Creates a new cached value wrapper.
+   * @param {function(): T} calculate the function to produce the value.
+   */
+  constructor(private readonly calculate: () => T) {}
+}
+
+/**
+ * A comparision entry holding an element and the element's set membership.
+ * @template T the element type.
+ */
+interface ComparisonEntry<T> {
+  /**
+   * The comparison element.
+   * @type {T}
+   * */
+  readonly element: T;
+  /**
+   * A boolean indicating whether the element is in the left set.
+   * @type {boolean}
+   */
+  readonly left: boolean;
+  /**
+   * A boolean indicating whether the element is in the right set.
+   * @type {boolean}
+   */
+  readonly right: boolean;
+}
+
+/**
+ * The result of a set comparison of two inputs.
+ * @template T the element type.
+ */
+class SetComparison<T> {
+  /**
+   * All entries from the left and right inputs.
+   * @type {ComparisonEntry[]}
+   */
+  readonly entries: ComparisonEntry<T>[];
+
+  private cachedLeft = new Lazy<T[]>(() =>
+    this.entries
+      .filter(entry => entry.left && !entry.right)
+      .map(entry => entry.element)
+  );
+  private cachedRight = new Lazy<T[]>(() =>
+    this.entries
+      .filter(entry => entry.right && !entry.left)
+      .map(entry => entry.element)
+  );
+  private cachedBoth = new Lazy<T[]>(() =>
+    this.entries
+      .filter(entry => entry.left && entry.right)
+      .map(entry => entry.element)
+  );
+
+  /**
+   * The elements that are only in the left input.
+   * @type {T}
+   */
+  get left() {
+    return this.cachedLeft.value;
+  }
+  /**
+   * The elements that are only in the right input.
+   * @type {T}
+   */
+  get right() {
+    return this.cachedRight.value;
+  }
+  /**
+   * The set intersection of the inputs.
+   * @type {T}
+   */
+  get both() {
+    return this.cachedBoth.value;
+  }
+
+  /**
+   * Adds an element and it's set membership to this comparison.
+   * @param {T} element the element.
+   * @param {boolean} left whether this element belongs to the left input.
+   * @param {boolean} right whether this element belongs to the right input.
+   */
+  private addElement(element: T, left: boolean, right: boolean) {
```

Comment:
Merge with `addEntry()` as discussed

Reply:
Removed all helper functions since their logic can be replace with a direct push.